homeassistant/components/geniushub/switch.py

```diff
@@ -0,0 +1,51 @@
+"""Support for Genius Hub switch/outlet devices."""
+from homeassistant.components.switch import SwitchDevice
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from . import DOMAIN, GeniusZone
+
+GH_ZONE_ATTR = "on / off"
+
+
+async def async_setup_platform(
+    hass: HomeAssistantType, config: ConfigType, async_add_entities, discovery_info=None
+) -> None:
+    """Set up the Genius Hub switch entities."""
+    if discovery_info is None:
+        return
+
+    broker = hass.data[DOMAIN]["broker"]
+
+    switches = [
+        GeniusSwitch(broker, z)
+        for z in broker.client.zone_objs
+        if GH_ZONE_ATTR in z.data["type"]
+    ]
+
+    async_add_entities(switches, update_before_add=True)
+
+
+class GeniusSwitch(GeniusZone, SwitchDevice):
+    """Representation of a Genius Hub switch."""
+
+    @property
+    def is_on(self) -> bool:
+        """Return the status of the sensor."""
+        # off, override, timer, footprint
+        # technially this could be untrue, if we're using the timer
+        return self._zone.data["mode"] != "off"
+
+    async def async_turn_off(self, **kwargs) -> None:
+        """Turn the switch off."""
+        await self._zone.set_mode("off")
+        # State is set optimistically in the command above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state()
```

Comment:
My understanding is that this is not required.

Reply:
I'm not sure what you mean by this (and the following comment re set_mode), not being required - these are the two methods by which we switch the zone on/off.

As far as I can tell (from the v1 api docs anyway), we can only switch zones on/off, not individual devices, which is why I implemented it as a zone..

Can you clarify please?

Reply:
Specifically, `async_schedule_update_ha_state` is not required.

Also, use `HA_HVAC_TO_GH`: have a look at, say a similar method in **climate.py**::
```python
    async def async_set_hvac_mode(self, hvac_mode: str) -> None:
        """Set a new hvac mode."""
        await self._zone.set_mode(HA_HVAC_TO_GH.get(hvac_mode))
```

Reply:
Ah! Ok, what does the update_ha_state bit do ? I assumed it was updating the home assistant ui or similar..)

Reply:
```python
 from homeassistant.const import (
    STATE_ON,
...

    self._zone.set_mode(HA_HVAC_TO_GH.get(STATE_OFF))
```

Reply:
> Ah! Ok, what does the update_ha_state bit do ? I assumed it was updating the home assistant ui or similar..)

HA does this automatically, since you did something that made it call the method.

Reply:
Hmm, I saw the HVAC -> GH conversions in climate.py, but figured that since I'm not converting from user input (switch ui only has an on/off switch) to internal values, using them would be a bit over the top.. 

Long term it would be handy if the outlets could also set the override time, though I assume this would involve creating something that isn't quite a pure SwitchDevice on the HA end?

Reply:
> would be a bit over the top

Yes, you're right, sorry, I was being pedantic.

> Long term it would be handy if the outlets could also set the override time

Changing the HA UI would take a lot of energy (you start by raising an issue in the architecture repo), however, it can be done for Automatons, etc., with service calls.

It is in my long-term plans to implement these (and other features), but I just don't have the spare time...